8281505: Add CompileCommand PrintIdealPhase

neliasso opened the pull request:
1. This patch adds the CompileCommand PrintIdealPhase. It allows to specify a list of compiler phases that will print the ideal IR.

-XX:CompileCommand=PrintIdealPhase,Class::method,MATCHING,FINAL_CODE"

The motivation for this is that during work with the IR framework I have found the need for printing specific phases, and the previously added PrintIdealLevel is forced to use levels that just prints to much. (I will remove PrintIdealLevel in a separate patch since it's not needed anymore.)

I have chosen not to add PrintIdealPhase as a VM flag, but keep it as CompileCommand/CompilerControl only.

2. This patch also changes the print_ideal output to emit the start tag with the phase name instead of the description, so that the name in the CompileCommand matches with what ends up in the tag. That tag is a new addition since PrintIdealLevel, so no one is using it yet.

3. This patch also removes an unused allocation in register_jfr_phasetype_serializer. 

Please review,
Nils Eliasson

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281505](https://bugs.openjdk.java.net/browse/JDK-8281505): Add CompileCommand PrintIdealPhase


### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 1a73bbfe5352bbb6b6b7503c160b247fc022df5d
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7392/head:pull/7392` \
`$ git checkout pull/7392`

Update a local copy of the PR: \
`$ git checkout pull/7392` \
`$ git pull https://git.openjdk.java.net/jdk pull/7392/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7392`

View PR using the GUI difftool: \
`$ git pr show -t 7392`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7392.diff">https://git.openjdk.java.net/jdk/pull/7392.diff</a>

</details>

vnkozlov reviewed (approved):
Good.

TobiHartmann commented on test/hotspot/jtreg/compiler/oracle/PrintIdealPhaseTest.java:
```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
```
```suggestion
 * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
```

TobiHartmann reviewed (approved):
Looks good to me.

neliasso commented:
@vnkozlov @TobiHartmann Thanks for having a look!

I updated the copyright year.

I also found that sometimes the test method can be compiled multiple times, so I changed the List to a HashSet. Also removed redundant command line parameters.

Best regards,
Nils

TobiHartmann reviewed (approved):
Looks good.

chhagedorn reviewed (approved):
Nice work! That will make things easier (and better in terms of performance) for [JDK-8280378](https://bugs.openjdk.java.net/browse/JDK-8280378). Looks good to me, too.

vnkozlov reviewed (approved):
Update is good.

vnkozlov reviewed (changes requested):
Looks like `zero` and `minimal` builds failed with this change:
```
src/hotspot/share/compiler/compilerDirectives.cpp:388:20: error: 'PrintIdealPhaseIndex' was not declared in this scope
[19](https://github.com/neliasso/jdk/runs/5155280503?check_suite_focus=true#step:19:19)
  388 |     if (!_modified[PrintIdealPhaseIndex]) {
[20](https://github.com/neliasso/jdk/runs/5155280503?check_suite_focus=true#step:19:20)
      |                    ^~~~~~~~~~~~~~~~~~~~
 ```

vnkozlov reviewed (approved):
Good.

Wait result of GIT testing.

neliasso commented:
Thanks for the reviews and feedback! @vnkozlov @TobiHartmann @chhagedorn

neliasso commented:
/integrate
